File: Sources/Usage/AntigravityUsageStore.swift
import Foundation

/// Gemini's slice of the panel: the last fetched quota snapshot plus the
/// account identity from oauth_creds.json. Timer-free like GrokUsageStore —
/// it rides `UsageStore.refresh()`'s cadence via `kickRefresh()` behind the
/// same 120s attempt floor, with a UserDefaults snapshot cache so relaunch
/// doesn't blank the strip.
@MainActor
final class AntigravityUsageStore: ObservableObject {
    static let shared = AntigravityUsageStore()

    @Published private(set) var snapshot: AntigravityQuotaSnapshot?
    /// Non-nil while the latest fetch ended in anything but data. Values in
    /// `snapshot` are the preserved last-good numbers in that case.
    @Published private(set) var statusCaption: String?
    @Published private(set) var lastUpdated: Date?
    @Published private(set) var accountEmail: String?
    @Published private(set) var loading = false

    /// Detection is launch-static, same as the other providers.
    let detection: AntigravityAuthDetection

    private var lastAttempt: Date?
    private static let cacheKey = "AntigravityUsageStore.lastSnapshot.v2"
    private static let cacheMaxAge: TimeInterval = 24 * 60 * 60
    private static let minAttemptGap: TimeInterval = 120

    private struct CachedSnapshot: Codable {
        var snapshot: AntigravityQuotaSnapshot
        var updatedAt: Date
    }

    private init() {
        if AppEnvironment.isDemo {
            if AppEnvironment.demoGuestFixturesEnabled {
                let now = Date()
                detection = .signedIn
                snapshot = AntigravityQuotaSnapshot(
                    buckets: [
                        AntigravityQuotaBucket(
                            bucketId: "gemini-weekly",
                            groupLabel: "Gemini Models",
                            window: "weekly",
                            usedPercent: 0.43,
                            resetAt: now.addingTimeInterval(4 * 24 * 3600)
                        ),
                        AntigravityQuotaBucket(
                            bucketId: "3p-weekly",
                            groupLabel: "Claude and GPT models",
                            window: "weekly",
                            usedPercent: 0.18,
                            resetAt: now.addingTimeInterval(4 * 24 * 3600)
                        ),
                    ],
                    tierID: "free-tier",
                    tierLabel: "Antigravity Starter Quota"
                )
                lastUpdated = now
            } else {
                detection = .notInstalled
            }
            return
        }
        detection = AntigravityCredentials.detect()
        guard detection == .signedIn else { return }
        loadIdentity()
        guard let data = UserDefaults.standard.data(forKey: Self.cacheKey),
              let cached = try? JSONDecoder().decode(CachedSnapshot.self, from: data),
              Date().timeIntervalSince(cached.updatedAt) <= Self.cacheMaxAge else { return }
        snapshot = cached.snapshot
        lastUpdated = cached.updatedAt
    }

    /// Tier chip for the Settings row / strip — compacted to sit beside
    /// PRO and MAX ("ANTIGRAVITY STARTER QUOTA" → "STARTER").
    var tierBadge: String? {
        AntigravityQuotaParser.compactTierBadge(
            label: snapshot?.tierLabel, tierID: snapshot?.tierID)
    }

    func kickRefresh() {
        guard !AppEnvironment.isDemo,
              detection == .signedIn,
              ProviderVisibilityStore.shared.antigravityPanelShown,
              !loading else { return }
        if let last = lastAttempt, Date().timeIntervalSince(last) < Self.minAttemptGap { return }
        lastAttempt = Date()
        loading = true
        Task { [weak self] in
            let outcome = await AntigravityUsageFetcher.fetch()
            self?.apply(outcome)
        }
    }

    private func apply(_ outcome: AntigravityUsageFetcher.Outcome) {
        loading = false
        switch outcome {
        case .success(let fresh):
            snapshot = fresh
            statusCaption = nil
            lastUpdated = Date()
            loadIdentity()
            persist(fresh)
        case .notRunning:
            // Antigravity keeps its quota in-process; with it closed there is
            // nothing to read anywhere on this machine. Keep the last good
            // numbers — the row's sync age already dates them, and the
            // warning line this used to add both nagged and leaked a ghost
            // "no data" tile through secondaryMissing (owner report,
            // 2026-08-09). Only a truly empty state earns a caption.
            statusCaption = snapshot == nil
                ? L10n.tr("start Antigravity to read quota")
                : nil
        case .failed(let message):
            // Keep the last good numbers; the caption admits staleness.
            statusCaption = message
        case .notInstalled:
            snapshot = nil
            statusCaption = nil
        }
    }

    /// The IDE roots write oauth_creds.json; the CLI keeps its token in the
    /// keychain and writes no such file, so the email comes off the language
    /// server instead. Reading the keychain would work too, but only by
    /// asking for the secret itself — which raises the keychain dialog on
    /// every refresh, for a hover-card caption.
    private func loadIdentity() {
        if let creds = AntigravityCredentials.loadCreds(from: AntigravityCredentials.credsURL()) {
            accountEmail = creds.email
            return
        }
        Task { [weak self] in
            let email = await AntigravityUsageFetcher.accountEmail()
            guard let self, let email else { return }
            self.accountEmail = email
        }
    }

    private func persist(_ fresh: AntigravityQuotaSnapshot) {
        let cached = CachedSnapshot(snapshot: fresh, updatedAt: Date())
        guard let data = try? JSONEncoder().encode(cached) else { return }
        UserDefaults.standard.set(data, forKey: Self.cacheKey)
    }
}

File: windows/src/AgentIsland/Usage/AntigravityUsageStore.cs
using System.ComponentModel;
using System.Windows.Threading;
using AgentIsland.Core;
using AgentIsland.Localization;
using AgentIsland.Model;

namespace AgentIsland.Usage;

/// Antigravity's slice of the panel: the last quota snapshot read from its
/// local language server, plus account identity. Timer-free — it rides
/// UsageStore.Refresh()'s cadence via KickRefresh() behind a 120s attempt
/// floor, with a Preferences snapshot cache so relaunch (or Antigravity not
/// running, the only time its quota is unreadable) doesn't blank the strip.
public sealed class AntigravityUsageStore : INotifyPropertyChanged
{
    public static AntigravityUsageStore Shared { get; } = new();

    private const string CacheKey = "AntigravityUsageStore.lastSnapshot.v1";
    private static readonly TimeSpan CacheMaxAge = TimeSpan.FromDays(8);
    private static readonly TimeSpan MinAttemptGap = TimeSpan.FromSeconds(120);

    private AntigravityQuotaSnapshot? _snapshot;
    private string? _statusCaption;
    private DateTimeOffset? _lastUpdated;
    private string? _accountEmail;
    private bool _loading;
    private DateTimeOffset? _lastAttempt;

    public event PropertyChangedEventHandler? PropertyChanged;

    /// Detection is launch-static, same as the other providers.
    public bool Detected { get; }

    private AntigravityUsageStore()
    {
        if (AppEnvironment.IsDemo)
        {
            if (!DemoGuestFixturesEnabled)
            {
                Detected = false;
                return;
            }
            var now = DateTimeOffset.Now;
            Detected = true;
            _snapshot = new AntigravityQuotaSnapshot(
                new[]
                {
                    new AntigravityQuotaBucket(
                        "gemini-weekly", "Gemini Models", "weekly", 0.37, now.AddDays(4.2)),
                },
                "free-tier",
                "Antigravity Starter Quota");
            _lastUpdated = now;
            return;
        }

        Detected = AntigravityCredentials.Detected;
        if (!Detected) return;
        if (Preferences.Get<CachedSnapshot?>(CacheKey) is not { } cached) return;
        if (cached.Snapshot is not { } stored) return;
        if (DateTimeOffset.Now - cached.UpdatedAt > CacheMaxAge) return;
        _snapshot = stored;
        _lastUpdated = cached.UpdatedAt;
        _accountEmail = cached.Email;
    }

    public AntigravityQuotaSnapshot? Snapshot
    {
        get => _snapshot;
        private set { _snapshot = value; Raise(nameof(Snapshot)); Raise(nameof(TierBadge)); }
    }

    /// Non-null while the latest fetch ended in anything but data. Values in
    /// Snapshot are the preserved last-good numbers in that case.
    public string? StatusCaption
    {
        get => _statusCaption;
        private set { _statusCaption = value; Raise(nameof(StatusCaption)); }
    }

    public DateTimeOffset? LastUpdated
    {
        get => _lastUpdated;
        private set { _lastUpdated = value; Raise(nameof(LastUpdated)); }
    }

    public string? AccountEmail
    {
        get => _accountEmail;
        private set { _accountEmail = value; Raise(nameof(AccountEmail)); }
    }

    public bool Loading
    {
        get => _loading;
        private set { _loading = value; Raise(nameof(Loading)); }
    }

    /// Tier chip for the Settings row / strip. Google's tier names are
    /// sentences ("Antigravity Starter Quota"); the compactor keeps the part
    /// that identifies the plan ("STARTER").
    public string? TierBadge =>
        AntigravityQuotaParser.CompactTierBadge(Snapshot?.TierLabel, Snapshot?.TierId);

    public void KickRefresh()
    {
        if (AppEnvironment.IsDemo) return;
        if (!Detected) return;
        if (!ProviderVisibilityStore.Shared.AntigravityPanelShown) return;
        if (Loading) return;
        if (_lastAttempt is { } last && DateTimeOffset.Now - last < MinAttemptGap) return;

        _lastAttempt = DateTimeOffset.Now;
        Loading = true;
        var dispatcher = System.Windows.Application.Current?.Dispatcher ?? Dispatcher.CurrentDispatcher;
        _ = Task.Run(async () =>
        {
            AntigravityUsageFetcher.Outcome outcome;
            try
            {
                outcome = await AntigravityUsageFetcher.Fetch();
            }
            catch (Exception error)
            {
                // A faulted fetch must never leave Loading latched — that
                // would freeze the strip until the app is relaunched.
                outcome = new AntigravityUsageFetcher.Outcome.Failed(error.Message);
            }
            await dispatcher.BeginInvoke(() => Apply(outcome));
        });
    }

    private void Apply(AntigravityUsageFetcher.Outcome outcome)
    {
        Loading = false;
        switch (outcome)
        {
            case AntigravityUsageFetcher.Outcome.Success success:
                Snapshot = success.Snapshot;
                if (success.Email is { Length: > 0 } email) AccountEmail = email;
                StatusCaption = null;
                LastUpdated = DateTimeOffset.Now;
                Persist(success.Snapshot, AccountEmail);
                break;
            case AntigravityUsageFetcher.Outcome.NotRunning:
                // Not an error: the embedded server is the only quota
                // source. Keep the last good numbers; the caption explains
                // why they aren't moving.
                StatusCaption = Snapshot is null
                    ? L10n.Tr("Installed — run agy to sign in")
                    : null;
                break;
            case AntigravityUsageFetcher.Outcome.Failed failed:
                // Keep the last good numbers; the caption admits staleness.
                StatusCaption = failed.Message;
                break;
            case AntigravityUsageFetcher.Outcome.NotInstalled:
                Snapshot = null;
                StatusCaption = null;
                break;
        }
    }

    private static void Persist(AntigravityQuotaSnapshot fresh, string? email) =>
        Preferences.Set(CacheKey, new CachedSnapshot(fresh, DateTimeOffset.Now, email));

    /// The guest fixtures only exist for the recording rig; a plain demo run
    /// still shows the classic Claude/Codex island.
    private static bool DemoGuestFixturesEnabled =>
        AppEnvironment.IsDemo && Environment.GetEnvironmentVariable("AGENTISLAND_DEMO_GUESTS") == "1";

    private void Raise(string name) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    /// Cache envelope. Public so the JSON serializer binds it the same way
    /// it binds UsageCacheSnapshot.
    public sealed record CachedSnapshot(
        AntigravityQuotaSnapshot? Snapshot, DateTimeOffset UpdatedAt, string? Email = null);
}
